pipe_gaps/core/gap_detector.py

```diff
@@ -0,0 +1,48 @@
+"""This module encapsulates the gap detection core algorithm."""
+import logging
+from datetime import timedelta
+import ciso8601
+
+from rich.progress import track
+
+logger = logging.getLogger()
+
+
+THRESHOLD = timedelta(hours=12, minutes=0, seconds=0)
+
+
+def detect(messages: list[dict], threshold: timedelta = THRESHOLD) -> list[tuple[dict, dict]]:
+    """Detects time gaps between AIS position messages.
+
+    Currently takes about 4 seconds to process 10M messages (i7-1355U 5.0GHz).
+
+    Args:
+        messages: List of AIS messages.
+        threshold: Any gap whose (end-start) is less than this threshold is discarded.
+
+    Returns:
+        list: gaps as 2d-tuple with (start, end) messages.
+    """
+    n = len(messages)
+    logger.info("Amount of messages: {}".format(n))
+
+    logger.info("Sorting messages by timestamp...")
+    messages_sorted = sorted(messages, key=lambda x: x["timestamp"])
+
+    logger.info("Detecting gaps...")
+    gaps = zip(messages_sorted[:-1], messages_sorted[1:])
+    gaps = tuple(
+        gap for gap in track(gaps, total=n, description="Filtering gaps:")
```

Comment:
You might want to disable this via a parameter (the progress bar I mean).

Reply:
Done